hack/jenkins/test-flake-chart/compute_flake_rate.go

```diff
@@ -0,0 +1,224 @@
+		for test, flakeRate := range environmentSplit {
+			fmt.Printf("%s,%s,%.2f\n", environment, test, flakeRate*100)
+		}
+	}
+}
+
+type TestEntry struct {
+	name        string
+	environment string
+	date        time.Time
+	status      string
+}
+
+// Reads CSV `file` and consumes each line to be a single TestEntry.
+func ReadData(file io.Reader) []TestEntry {
+	testEntries := []TestEntry{}
+
+	fileReader := bufio.NewReaderSize(file, 256)
+	previousLine := []string{"", "", "", "", "", ""}
+	firstLine := true
+	for {
+		lineBytes, _, err := fileReader.ReadLine()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			exit("Error reading data CSV", err)
+		}
+		line := string(lineBytes)
+		fields := strings.Split(line, ",")
+		if firstLine {
+			if len(fields) != 6 {
+				exit(fmt.Sprintf("Data CSV in incorrect format. Expected 6 columns, but got %d", len(fields)), fmt.Errorf("bad CSV format"))
+			}
+			firstLine = false
+		}
+		for i, field := range fields {
+			if field == "" {
+				fields[i] = previousLine[i]
+			}
+		}
+		if len(fields) != 6 {
+			fmt.Printf("Found line with wrong number of columns. Expectd 6, but got %d - skipping\n", len(fields))
+			continue
+		}
+		previousLine = fields
+		if fields[4] == "Passed" || fields[4] == "Failed" {
+			date, err := time.Parse("2006-01-02", fields[1])
+			if err != nil {
+				fmt.Printf("Failed to parse date: %v\n", err)
+			}
+			testEntries = append(testEntries, TestEntry{
+				name:        fields[3],
+				environment: fields[2],
+				date:        date,
+				status:      fields[4],
+			})
+		}
+	}
+	return testEntries
+}
+
+// Splits `testEntries` up into maps indexed first by environment and then by test.
+func SplitData(testEntries []TestEntry) map[string]map[string][]TestEntry {
+	splitEntries := make(map[string]map[string][]TestEntry)
+
+	for _, entry := range testEntries {
+		appendEntry(splitEntries, entry.environment, entry.name, entry)
+	}
+
+	return splitEntries
+}
+
+// Appends `entry` to `splitEntries` at the `environment` and `test`.
+func appendEntry(splitEntries map[string]map[string][]TestEntry, environment, test string, entry TestEntry) {
+	// Lookup the environment.
+	environmentSplit, ok := splitEntries[environment]
+	if !ok {
+		// If the environment map is missing, make a map for this environment and store it.
+		environmentSplit = make(map[string][]TestEntry)
+		splitEntries[environment] = environmentSplit
+	}
+
+	// Lookup the test.
+	testSplit, ok := environmentSplit[test]
+	if !ok {
+		// If the test is missing, make a slice for this test.
+		testSplit = make([]TestEntry, 0)
+		// The slice is not inserted, since it will be replaced anyway.
+	}
+	environmentSplit[test] = append(testSplit, entry)
+}
+
+// Filters `splitEntries` to include only the most recent `date_range` dates.
+func FilterRecentEntries(splitEntries map[string]map[string][]TestEntry, dateRange uint) map[string]map[string][]TestEntry {
+	filteredEntries := make(map[string]map[string][]TestEntry)
```

Comment:
Would be good to make `map[string]map[string][]TestEntry` into a type

Reply:
Done.